aea/protocols/dialogue/base.py

```diff
@@ -857,66 +851,119 @@ def _basic_validation_non_initial_message(
                 ),
             )
 
-        last_message_id = self.last_message.message_id  # type: ignore
-        if message_id != last_message_id + 1:
-            return (
-                False,
-                "Invalid message_id. Expected {}. Found {}.".format(
-                    last_message_id + 1, message_id
-                ),
-            )
-        if target < 1:
-            return (
-                False,
-                "Invalid target. Expected a value greater than or equal to 1. Found {}.".format(
-                    target
-                ),
+        # it's non initial check, assume there some messages in the dialogue
+        if self.is_empty:
+            return False, "Can not reply cause no messages in the dialogue."
+
+        err = self._validate_message_id(message)
+        if err:
+            return False, err
+
+        err = self._validate_message_target(message)
+        if err:
+            return False, err
+
+        return True, "The non-initial message passes basic validation."
+
+    def _validate_message_target(self, message: Message) -> Optional[str]:
+        """Check message target corresponds to messages in the dialogue, if not return error string."""
+        target = message.target
+        performative = message.performative
+
+        if (
+            message.message_id == self.STARTING_MESSAGE_ID
+            and target != self.STARTING_TARGET
+        ):
+            return "Invalid target. Expected 0. Found {}.".format(target)
+
+        if (
+            message.message_id != self.STARTING_MESSAGE_ID
+            and target == self.STARTING_TARGET
+        ):
+            return "Invalid target. Expected a absolute value greater or less than 0. Found {}.".format(
+                target
             )
-        if last_message_id < target:
-            return (
-                False,
-                "Invalid target. Expected a value less than or equal to {}. Found {}.".format(
-                    last_message_id, target
-                ),
+
+        # quick target check.
+        latest_ids: List[int] = []
+
+        if self.last_incoming_message:
+            latest_ids.append(abs(self.last_incoming_message.message_id))
+
+        if self.last_outgoing_message:
+            latest_ids.append(abs(self.last_outgoing_message.message_id))
+
+        if abs(target) > max(latest_ids):
+            return "Invalid target. Expected a value less than or equal to abs({}). Found abs({}).".format(
+                self._last_message_id, target
             )
 
-        target_message = self._get_message(target)
+        # detailed target check
+        target_message = self.get_message_by_id(target)
+
+        if not target_message:
+            return "Invalid target {}. target_message can not be found.".format(target)
```

Comment:
Line 879 to 902 can be replaced with a simple call to `has_message_id(target)`.

Reply:
target message is used next, but updated get_message_by_id and has_mesasge_id to use without iterations